econml/sklearn_extensions.py

```diff
@@ -0,0 +1,975 @@
+            np.apply_along_axis(lambda s: norm.ppf(
+                lower / 100, scale=s), 0, sd_pred)
+        y_upper = y_pred + \
+            np.apply_along_axis(lambda s: norm.ppf(
+                upper / 100, scale=s), 0, sd_pred)
+        return y_lower, y_upper
+
+    def _get_coef_correction(self, X, y, y_pred, sample_weight, theta_hat):
+        # Assumes flattened y
+        n_samples, n_features = X.shape
+        y_res = np.ndarray.flatten(y) - y_pred
+        # Compute weighted residuals
+        if sample_weight is not None:
+            y_res_scaled = y_res * sample_weight / np.sum(sample_weight)
+        else:
+            y_res_scaled = y_res / n_samples
+        delta_coef = np.matmul(
+            np.matmul(theta_hat, X.T), y_res_scaled)
+        return delta_coef
+
+    def _get_optimal_alpha(self, alpha_grid, X, y, sample_weight):
+        # To be done once per target. Assumes y can be flattened.
+        cv_estimator = WeightedLassoCV(alphas=alpha_grid, cv=5, fit_intercept=self.fit_intercept)
+        cv_estimator.fit(X, y.flatten(), sample_weight=sample_weight)
+        return cv_estimator.alpha_
+
+    def _get_theta_hat(self, X, sample_weight):
+        # Assumes that X has already been offset
+        n_samples, n_features = X.shape
+        coefs = np.empty((n_features, n_features - 1))
+        tausq = np.empty(n_features)
+        # Compute Lasso coefficients for the columns of the design matrix
+        for i in range(n_features):
+            y = X[:, i]
+            X_reduced = X[:, list(range(i)) + list(range(i + 1, n_features))]
+            # Call weighted lasso on reduced design matrix
+            # Inherit some parameters from the parent
+            local_wlasso = WeightedLasso(
+                alpha=self.alpha,
+                fit_intercept=False,
+                max_iter=self.max_iter,
+                tol=self.tol
+            ).fit(X_reduced, y, sample_weight=sample_weight)
+            coefs[i] = local_wlasso.coef_
+            # Weighted tau
+            if sample_weight is not None:
+                y_weighted = y * sample_weight / np.sum(sample_weight)
+            else:
+                y_weighted = y / n_samples
+            tausq[i] = np.dot(y - local_wlasso.predict(X_reduced), y_weighted)
+        # Compute C_hat
+        C_hat = np.diag(np.ones(n_features))
+        C_hat[0][1:] = - coefs[0]
+        for i in range(1, n_features):
+            C_hat[i][:i] = - coefs[i][:i]
+            C_hat[i][i + 1:] = - coefs[i][i:]
+        # Compute theta_hat
+        theta_hat = np.matmul(np.diag(1 / tausq), C_hat)
+        return theta_hat
+
+    def _get_unscaled_coef_var(self, X, theta_hat, sample_weight):
+        if sample_weight is not None:
+            weights_mat = np.diag(sample_weight / np.sum(sample_weight))
+            sigma = np.matmul(X.T, np.matmul(weights_mat, X))
+        else:
+            sigma = np.matmul(X.T, X) / X.shape[0]
+        _unscaled_coef_var = np.matmul(
```

Comment:
Replace `np.matmul(np.matmul(X, Y), Z)` with equivalent `X @ Y @ Z`

I'd have provided this in a suggestion, but GitHub doesn't support multi-line suggestions yet.

Reply:
@kbattocchi This is super cool, I didn't know you could do this. Replacing this is all my code. Thanks for the detailed comments on the PR, I'm learning some new stuff :)